> Need to address the PR comments below, and fix the following errors:
> 
> ```
> error[E0277]: the trait bound `service::state::State: std::clone::Clone` is not satisfied
>    --> rama-fp/src/service/mod.rs:125:14
>     |
> 125 |             .service(match_service!{
>     |              ^^^^^^^ the trait `std::clone::Clone` is not implemented for `service::state::State`
>     |
>     = help: the trait `rama::service::Layer<S>` is implemented for `rama::service::layer::Stack<Inner, Outer>`
> ...
> ```
> 
> ```
> error[E0277]: the trait bound `rama::http::Body: std::clone::Clone` is not satisfied
>    --> rama-fp/src/service/mod.rs:125:14
>     |
> 125 |             .service(match_service!{
>     |              ^^^^^^^ the trait `std::clone::Clone` is not implemented for `rama::http::Body`
>     |
>     = help: the trait `rama::service::Layer<S>` is implemented for `rama::service::layer::Stack<Inner, Outer>`
>     = note: required for `rama::http::matcher::HttpMatcher<service::state::State, rama::http::Body>` to implement `std::clone::Clone`
> ...
> ```

Those errors are due to the fact that we have `#[derive(...,? Clone)]` statements on the `HttpMatcher(Kind)`. When you use derive Clone with generics you also automatically require the parameters to be `Cloneable`. As such when using generics it's usually better (and in this case even required) to implement Clone yourself.

See <https://github.com/dmackdev/rama/pull/1/files#r1545646890> and the rest of my review for more info.

--- a/src/http/matcher/mod.rs
+++ b/src/http/matcher/mod.rs
@@ -5,6 +5,13 @@
 //! [`service::Matcher`]: crate::service::Matcher
 //! [`http::Request`]: crate::http::Request
 //! [`service::matcher` module]: crate::service::matcher
+use crate::{
+    http::Request,
+    service::{context::Extensions, matcher::IteratorMatcherExt, Context},
+    stream::matcher::SocketMatcher,
+};
+use std::fmt;
+use std::sync::Arc;
 
 mod method;
 #[doc(inline)]
@@ -28,15 +35,6 @@ pub use path::{PathMatcher, UriParams, UriParamsDeserializeError};
 mod header;
 #[doc(inline)]
 pub use header::HeaderMatcher;
-
-use crate::{
-    http::Request,
-    service::{context::Extensions, matcher::IteratorMatcherExt, Context},
-    stream::matcher::SocketMatcher,
-};
-
-use core::fmt;
-use std::sync::Arc;
 
 #[derive(Debug, Clone)]
 /// A matcher that is used to match an http [`Request`]

--- a/src/http/matcher/mod.rs
+++ b/src/http/matcher/mod.rs
@@ -36,14 +36,22 @@ mod header;
 #[doc(inline)]
 pub use header::HeaderMatcher;
 
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 /// A matcher that is used to match an http [`Request`]
 pub struct HttpMatcher<State, Body> {
     kind: HttpMatcherKind<State, Body>,
     negate: bool,
 }
 
-#[derive(Clone)]
+impl<State, Body> Clone for HttpMatcher<State, Body> {
+    fn clone(&self) -> Self {
+        Self {
+            kind: self.kind.clone(),
+            negate: self.negate,
+        }
+    }
+}
+
 /// A matcher that is used to match an http [`Request`]
 pub enum HttpMatcherKind<State, Body> {
     /// zero or more [`HttpMatcher`]s that all need to match in order for the matcher to return `true`.
@@ -71,21 +79,41 @@ pub enum HttpMatcherKind<State, Body> {
     Custom(Arc<dyn crate::service::Matcher<State, Request<Body>>>),
 }
 
-impl<State, Body> fmt::Debug for HttpMatcherKind<State, Body> {
+impl<State, Body> Clone for HttpMatcherKind<State, Body> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::All(inner) => Self::All(inner.clone()),
+            Self::Method(inner) => Self::Method(*inner),
+            Self::Path(inner) => Self::Path(inner.clone()),
+            Self::Domain(inner) => Self::Domain(inner.clone()),
+            Self::Version(inner) => Self::Version(*inner),
+            Self::Any(inner) => Self::Any(inner.clone()),
+            Self::Uri(inner) => Self::Uri(inner.clone()),
+            Self::Header(inner) => Self::Header(inner.clone()),
+            Self::Socket(inner) => Self::Socket(inner.clone()),
+            Self::Custom(inner) => Self::Custom(inner.clone()),
+        }
+    }
+}
+
+impl<State, Body> fmt::Debug for HttpMatcherKind<State, Body>
+where
+    State: fmt::Debug,
+    Body: fmt::Debug,
+{
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        // match self {
-        //     Self::All(inner) => f.debug_tuple("All").field(inner).finish(),
-        //     Self::Method(inner) => f.debug_tuple("Method").field(inner).finish(),
-        //     Self::Path(inner) => f.debug_tuple("Path").field(inner).finish(),
-        //     Self::Domain(inner) => f.debug_tuple("Domain").field(inner).finish(),
-        //     Self::Version(inner) => f.debug_tuple("Version").field(inner).finish(),
-        //     Self::Any(inner) => f.debug_tuple("Any").field(inner).finish(),
-        //     Self::Uri(inner) => f.debug_tuple("Uri").field(inner).finish(),
-        //     Self::Header(inner) => f.debug_tuple("Header").field(inner).finish(),
-        //     Self::Socket(inner) => f.debug_tuple("Socket").field(inner).finish(),
-        //     Self::Custom(_) => f.debug_tuple("Custom").finish(),
-        // }
-        todo!("The type parameters State and Body do not implement fmt::Debug, so cannot debug format inner fields.")
+        match self {
+            Self::All(inner) => f.debug_tuple("All").field(inner).finish(),
+            Self::Method(inner) => f.debug_tuple("Method").field(inner).finish(),
+            Self::Path(inner) => f.debug_tuple("Path").field(inner).finish(),
+            Self::Domain(inner) => f.debug_tuple("Domain").field(inner).finish(),
+            Self::Version(inner) => f.debug_tuple("Version").field(inner).finish(),
+            Self::Any(inner) => f.debug_tuple("Any").field(inner).finish(),
+            Self::Uri(inner) => f.debug_tuple("Uri").field(inner).finish(),
+            Self::Header(inner) => f.debug_tuple("Header").field(inner).finish(),
+            Self::Socket(inner) => f.debug_tuple("Socket").field(inner).finish(),
+            Self::Custom(_) => f.debug_tuple("Custom").finish(),
+        }
     }
 }
 

--- a/src/http/matcher/mod.rs
+++ b/src/http/matcher/mod.rs
@@ -547,6 +547,39 @@ impl<State, Body> HttpMatcher<State, Body> {
         Self::method_get().and_path(path)
     }
 
+    /// Create a matcher that matches according to a custom predicate.
+    ///
+    /// See [`crate::service::Matcher`] for more information.
+    pub fn custom<M>(matcher: M) -> Self
+    where
+        M: crate::service::Matcher<State, Request<Body>>,
+    {
+        Self {
+            kind: HttpMatcherKind::Custom(Arc::new(matcher)),
+            negate: false,
+        }
+    }
+
+    /// Add a custom matcher to match on top of the existing set of [`HttpMatcher`] matchers.
+    ///
+    /// See [`crate::service::Matcher`] for more information.
+    pub fn and_custom<M>(self, matcher: M) -> Self
+    where
+        M: crate::service::Matcher<State, Request<Body>>,
+    {
+        self.and(Self::custom(matcher))
+    }
+
+    /// Create a custom matcher to match as an alternative to the existing set of [`HttpMatcher`] matchers.
+    ///
+    /// See [`crate::service::Matcher`] for more information.
+    pub fn or_custom<M>(self, matcher: M) -> Self
+    where
+        M: crate::service::Matcher<State, Request<Body>>,
+    {
+        self.or(Self::custom(matcher))
+    }
+
     /// Create a [`PathMatcher`] matcher to match for a POST request.
     pub fn post(path: impl AsRef<str>) -> Self {
         Self::method_post().and_path(path)

